pkg/metrics/statsd/statsd_test.go

```diff
@@ -83,3 +86,247 @@ func TestEmitFloat(t *testing.T) {
+	env.EXPECT().Hostname().AnyTimes().Return("test-host")
+
+	socket := "127.0.0.1:8001"
+
+	os.Setenv(statsdSocketEnv, "udp:"+socket)
+	c2, _ := net.ListenPacket("udp", socket)
+
+	s := &statsd{
+		env:       env,
+		account:   "*",
+		namespace: "*",
+		ch:        make(chan *metric),
+		now:       func() time.Time { return time.Time{} },
+	}
+
+	c := make(chan string)
+	go func() {
+		buf := make([]byte, 1024)
+		// set 5 second read timeout. That should be plenty of time to receive the emitted Gauge
+		c2.SetReadDeadline(time.Now().Add(5 * time.Second))
+		m := ""
+		n, _, err := c2.ReadFrom(buf)
+		if err != nil {
+			m = err.Error()
+		} else {
+			m = string(buf[:n])
+		}
+
+		c <- m
+	}()
+
+	go s.run()
+	s.EmitGauge("tests.test_key", 42, map[string]string{"key": "value"})
+	m := <-c
+	if m != `{"Metric":"tests.test_key","Account":"*","Namespace":"*","Dims":{"hostname":"test-host","key":"value","location":"eastus"},"TS":"0001-01-01T00:00:00.000"}:42|g`+"\n" {
+		t.Error(errors.New(m))
+	}
+}
+
+func TestGetConnectionDetails(t *testing.T) {
+	controller := gomock.NewController(t)
+	defer controller.Finish()
+
+	for _, tt := range []struct {
+		name         string
+		set          bool
+		isLocalDev   bool
+		socketstring string
+		network      string
+		address      string
+		wantError    bool
+	}{
+		{
+			name:      "Old Bevaviour / production mode",
+			set:       false,
+			network:   "unix",
+			address:   "/var/etw/mdm_statsd.socket",
+			wantError: false,
+		},
+		{
+			name:       "Old Bevaviour / localdev mode",
+			set:        false,
+			isLocalDev: true,
+			network:    "unix",
+			address:    "mdm_statsd.socket",
+			wantError:  false,
+		},
+		{
+			name:         "Valid UDP env variable",
+			set:          true,
+			socketstring: "udp:127.0.0.1:9000",
+			network:      "udp",
+			address:      "127.0.0.1:9000",
+			wantError:    false,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			env := mock_env.NewMockInterface(controller)
+
+			if !tt.set {
+				if tt.isLocalDev {
+					env.EXPECT().IsLocalDevelopmentMode().Return(true)
+				} else {
+					env.EXPECT().IsLocalDevelopmentMode().Return(false)
+				}
+			}
+
+			s := &statsd{
+				env: env,
+			}
+
+			os.Unsetenv(statsdSocketEnv)
+			if tt.set {
+				os.Setenv(statsdSocketEnv, tt.socketstring)
+			}
+
+			network, address, err := s.getConnectionDetails()
+			if err != nil {
+				fmt.Println(err.Error())
+			}
+
+			if tt.wantError {
+				if err == nil {
+					t.Fail()
+				}
+			} else {
+				if err != nil && tt.wantError == false {
+					fmt.Println(err.Error())
+					t.Fail()
+				}
+				if network != tt.network {
+					t.Fail()
+				}
+				if address != tt.address {
+					t.Fail()
+				}
+			}
+		})
+	}
+}
+
+func TestParseSocketEnv(t *testing.T) {
+	controller := gomock.NewController(t)
+	defer controller.Finish()
+
+	for _, tt := range []struct {
+		name      string
+		socketEnv string
+		network   string
+		address   string
+		wantError bool
```

Comment:
I think this needs to be a string with and we must check for a specific error. Given that we return multiple errors we need to make sure that we are hitting the case we intend the case to hit.

Similar to this:

https://github.com/Azure/ARO-RP/blob/1ae6115606b60c4f530ae4154cae8f734e161cc2/pkg/api/admin/openshiftcluster_validatestatic_test.go#L58-L65

Reply:
Changed to strings and specific errors